src/util/threadpool.rs

```diff
@@ -0,0 +1,610 @@
+//    `group_concurrency_on_busy`.
+// 2. If more than one groups meet the condition 1, running the one who
+//    comes first.
+// If no group's running number is smaller than `group_concurrency_on_busy`,
+// choose according to the following rules:
+// 1. Choose the groups with least running tasks.
+// 2. If more than one groups has the least running tasks,
+//    choose the one whose task comes first(with the minum task's ID)
+pub struct BigGroupThrottledQueue<T> {
+    // tasks in pending. tasks in `pending_tasks` have higher priority than
+    // tasks in waiting_queue.
+    pending_tasks: BinaryHeap<Task<T>>,
+    // group_id => tasks array. If `group_concurrency[group_id]` is bigger than
+    // `group_concurrency_on_busy`(which means there may more than
+    // `group_concurrency_on_busy` tasks of the group are running), the rest
+    // of the group's tasks would be pushed into `waiting_queue[group_id]`
+    waiting_queue: HashMap<T, VecDeque<Task<T>>>,
+    // group_id => running_num+pending num. It means there may
+    // `group_concurrency[group_id]` tasks of the group are running.
+    group_concurrency: HashMap<T, usize>,
+    // max num of threads each group can run on when pool is busy.
+    // each value in group_concurrency shouldn't bigger than this value.
+    group_concurrency_on_busy: usize,
+}
+
+impl<T: Hash + Eq + Send + Clone> BigGroupThrottledQueue<T> {
+    pub fn new(group_concurrency_on_busy: usize) -> BigGroupThrottledQueue<T> {
+        BigGroupThrottledQueue {
+            group_concurrency: HashMap::default(),
+            waiting_queue: HashMap::default(),
+            pending_tasks: BinaryHeap::new(),
+            group_concurrency_on_busy: group_concurrency_on_busy,
+        }
+    }
+
+    // Try push into pending. Return none on success,return Some(task) on failed.
+    #[inline]
+    fn try_push_into_pending(&mut self, task: Task<T>) -> Option<Task<T>> {
+        let count = self.group_concurrency.entry(task.group_id.clone()).or_insert(0);
+        if *count >= self.group_concurrency_on_busy {
+            return Some(task);
+        }
+        *count += 1;
+        self.pending_tasks.push(task);
+        None
+    }
+
+    #[inline]
+    fn pop_task_from_waiting_queue(&mut self) -> Option<Task<T>> {
+        let group_id = self.pop_group_id_from_waiting_queue();
+        if group_id.is_none() {
+            return None;
+        }
+        let group_id = group_id.unwrap();
+        let task = self.pop_from_waiting_queue_with_group_id(&group_id);
+        // update group_concurrency since current task is going to run.
+        let mut count = self.group_concurrency.entry(group_id.clone()).or_insert(0 as usize);
+        *count += 1;
+        Some(task)
+    }
+
+    #[inline]
+    fn pop_from_waiting_queue_with_group_id(&mut self, group_id: &T) -> Task<T> {
+        let (left_tasks_num, task) = {
+            let mut waiting_tasks = self.waiting_queue.get_mut(group_id).unwrap();
```

Comment:
This shouldn't happen since we check waiting_queue with `group_id` before any  `pop_from_waiting_queue_with_group_id ` happens. @andelf